src/arch/riscv/linux/se_workload.cc

```diff
@@ -134,6 +136,381 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
+        break;
+    case RiscvLinux::Marchid:
+        pair->value = tc->readMiscRegNoEffect(CSRData.at(CSR_MARCHID).physIndex);
+        break;
+    case RiscvLinux::Mimpid:
+        pair->value = tc->readMiscRegNoEffect(CSRData.at(CSR_MIMPID).physIndex);
+        break;
+    case RiscvLinux::BaseBehavior:
+        {
+            MISA misa = tc->readMiscRegNoEffect(MISCREG_ISA);
+            RiscvLinux::key_base_behavior_t *base_behavior =
+                (RiscvLinux::key_base_behavior_t *)&pair->value;
+            if (misa.rvi && misa.rvm && misa.rva) {
+                base_behavior->ima = 1;
+            }
+        }
+        break;
+    case RiscvLinux::IMAExt0:
+        {
+            MISA misa = tc->readMiscRegNoEffect(MISCREG_ISA);
+            RiscvLinux::key_ima_ext_0_t *ext = (RiscvLinux::key_ima_ext_0_t *)&pair->value;
+            if (misa.rvf && misa.rvd) ext->FD = 1;
+            if (misa.rvc) ext->C = 1;
+            if (misa.rvv) ext->V = 1;
+            ext->ZBA = 1;
+            ext->ZBB = 1;
+            ext->ZBS = 1;
+            ext->ZICBOZ = 1;
+            ext->ZBC = 1;
+            ext->ZBKB = 1;
+            ext->ZBC = 1;
+            ext->ZBKX = 1;
+            ext->ZKND = 1;
+            ext->ZKNE = 1;
+            ext->ZKNH = 1;
+            ext->ZKSED = 1;
+            ext->ZKSH = 1;
+            ext->ZKT = 1;
+            ext->ZFH = 1;
+            ext->ZFHMIN = 1;
+            ext->ZVFH = 1;
+            ext->ZVFHMIN = 1;
+            ext->ZICOND = 1;
+            ext->ZVE64D = 1;
+            ext->ZCB = 1;
+            ext->ZCD = 1;
+            ext->ZCF = 1;
+        }
+        break;
+    case RiscvLinux::Cpuperf0:
+    case RiscvLinux::MisalignedScalarPerf:
+        pair->value = RiscvLinux::Slow;
+        break;
+    case RiscvLinux::ZicbozBlockSize:
+        pair->value = tc->getSystemPtr()->cacheLineSize();
+        break;
+    case RiscvLinux::HighestVirtAddress:
+        pair->value = tc->getProcessPtr()->memState->getMmapEnd();
+        break;
+
+    /*
+     * For forward compatibility, unknown keys don't fail the whole
+     * call, but get their element key set to -1 and value set to 0
+     * indicating they're unrecognized.
+     */
+    default:
+        pair->key = -1;
+        pair->value = 0;
+        break;
+    }
+}
+
+template <class OS>
+static int
+hwprobe_get_values(ThreadContext *tc, VPtr<> pairs, typename OS::size_t pair_count,
+                   typename OS::size_t cpusetsize, VPtr<> cpus_user, unsigned int flags)
+{
+    /* Check the reserved flags. */
+    if (flags != 0)
+        return -EINVAL;
+
+    int ret = 0;
+    RiscvLinux::riscv_hwprobe *pair;
+    RiscvLinux::cpumask_t *cpu_online_mask = get_cpu_online_mask(tc);
+    RiscvLinux::cpumask_t *cpus = cpumask_malloc(tc);
+    BufferArg pairs_buf(pairs, sizeof(RiscvLinux::riscv_hwprobe) * pair_count);
+
+    if (cpu_online_mask == nullptr || cpus == nullptr) {
+        ret = -ENOMEM;
+        goto exit;
```

Comment:
I would strongly prefer not having "goto" and instead duplicating the code. It's only used in one or two places, so I don't think that it's worth the downsides.

Also, I strongly prefer *always* using `{ }` around if/while statements. I have had to fix multiple very difficult to find bugs because people want to save 2 characters 😉

Reply:
Re: `goto`: There's only three places in the entire gem5 codebase this is used. Let's not add another, if we can possibly avoid it.

Reply:
I've removed all `goto` and added `{ }`.